### PR TITLE
(MODULES-3035) Manage configuration settings

### PR DESCRIPTION
**Pull request by ferventcoder**
~~Requires #1 to be merged first.~~

Provide the ability to manage config settings with a custom type and
provider. This allows a config setting to be ensurable, and also
allows for a config settings to be manageable in the same converge
that Chocolatey is installed during.

Things to accomplish:

- [x] Initial code
- [x] Specs
- [x] Documentation
- [x] Reference tests
- [x] insync? when the name includes Password

### CODE REVIEW

File: lib/puppet/provider/chocolateyconfig/windows.rb [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+    if @property_flush[:ensure] != :absent
+      args << '--value' << resource[:value]
+    end
+
+    # not sure if we should do this or just punt over to rexml and delete the entire line for unsets
+    # may not be a bad idea entirely, as choco would just put back the default values automatically
+    # that way.
+    if @property_flush[:ensure] == :absent && choco_version < Gem::Version.new(CONFIG_MINIMUM_SUPPORTED_CHOCO_VERSION_UNSET)
+      #todo use xml writing instead
```
Comment (ferventcoder) [inline lib/puppet/provider/chocolateyconfig/windows.rb]: One of the things left to do.

File: lib/puppet/type/chocolateyconfig.rb [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    validate do |value|
+      if value.nil? or value.empty?
+        raise ArgumentError, "A non-empty value must be specified. To unset value, use ensure => absent"
+      end
+    end
+
+    def insync?(is)
+      #todo: If name contains password, it is
+      # always in sync if there is a value
```
Comment (ferventcoder) [inline lib/puppet/type/chocolateyconfig.rb]: If the name contains password, we need to state that it is insync if there is a value.

**Comment by ferventcoder** [comment]
@puppetlabs/windows this is ready to start reviewing.

File: lib/puppet/type/chocolateyconfig.rb [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+      if value.nil? or value.empty?
+        raise ArgumentError, "A non-empty value must be specified. To unset value, use ensure => absent"
+      end
+    end
+
+    def insync?(is)
+      if (self[:title] =~ /password/i)
+        # If name contains password, it is
+        # always in sync if there is a value
+        return (is.nil? || is.empty?) == (should.nil? || should.empty?)
```
Comment (ferventcoder) [inline lib/puppet/type/chocolateyconfig.rb]: Not really sure how to test this in specs. I have a pretty good idea of testing it in reference tests though.

**Comment by glennsarti** [comment]
Assuming adhoc build is ok, 👍  for merge

**Pull request merged**